### PR TITLE
Build dev container for arm64 and use inline caching

### PR DESCRIPTION
**Pull request by jgiannuzzi**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/jgiannuzzi/mlflow/pull/11406?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11406/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11406
```

</p>
</details>

### What changes are proposed in this pull request?

This PR enables building the dev container for `arm64`, enabling native support on Apple Silicon Macs. It builds the two architectures in parallel (`amd64` and `arm64`) and pushes them both as a unified image. The frontend is built on the native architecture of the GitHub runner, and the rest of the image is built under emulation for `arm64`. Finally, inline caching is enabled and should limit the time taken to build when no changes have happened.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

See the following test runs on my fork:
- [Initial run, no cache at all](https://github.com/jgiannuzzi/mlflow/actions/runs/9065448143)
- [Run after a minor change, cache gets used](https://github.com/jgiannuzzi/mlflow/actions/runs/9067339663)

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by jgiannuzzi** [comment]
Hi @harupy, you seem to be the main contributor to the devcontainer. Could you please review this PR?

**Comment by BenWilson2** [comment]
Hi @jgiannuzzi sorry for missing this. Can you rebase on master and push again? Tag me after you push and I'll trigger the full CI suite!

File: .github/workflows/devcontainer.yml [hunk 1]
```diff
@@ -24,26 +24,39 @@ defaults:
...
+        arch: [amd64, arm64]
     steps:
       - uses: actions/checkout@v4
       - name: Get image name
+        id: image
         run: |
           IMAGE_NAME=$(yq '.services.mlflow.image' .devcontainer/docker-compose.yml)
           echo "IMAGE_NAME=$IMAGE_NAME" >> $GITHUB_ENV
+          echo "name=$IMAGE_NAME" >> $GITHUB_OUTPUT
+      - name: Set up QEMU
```
Comment (BenWilson2) [inline .github/workflows/devcontainer.yml:45]: Great emulation idea!

**Review by BenWilson2 — APPROVED** [review]
Approach looks good to me - thanks for the caching approach and the arm emulation will definitely come in handy :) 
Once we get the rebase and CI all green, I'll merge!

**Comment by jgiannuzzi** [comment]
Thanks for the review @BenWilson2! I just rebased, tested locally, and pushed, so ready to run the CI :)

**Comment by harupy** [comment]
@jgiannuzzi The devcontainer workflow failed on master: https://github.com/mlflow/mlflow/actions/runs/9654229639/job/26628074221. would you mind taking a look?

**Comment by jgiannuzzi** [comment]
@harupy I _think_ this is a transient error (500 returned by ghcr). It should go away by retrying the failed jobs.

**Comment by jgiannuzzi** [comment]
@harupy the next commit triggered a successful run of the devcontainer workflow so I think it's all good now: https://github.com/mlflow/mlflow/actions/runs/9654980305

**Comment by harupy** [comment]
@jgiannuzzi thanks!

**Pull request merged**